### PR TITLE
[MRG]Dcgan

### LINKED ISSUE #129: ImageGenerator
### Suggested Name

ImageGenerator

### Task Description

Generating new images similar to the training images. The input data should be a set of images without labels.

### Evaluation Metrics

Human Evaluation

### Benchmark Datasets

CelebA

### Reason

NA

### Solution

Do a baseline solution first.
Do not search for different architectures.
Just use a default architecture, which only has dynamic parts in the input and output sizes.

### PR DESCRIPTION
**Pull request by tl-yang**
This pull request resolves #129 

The pull request is in progress.

The details of the pull request:

- Only for 32*32 images for now
- GANModelTrainer needs to integrate with the original ModelTrainer
- GAN Discriminator and Generator model need futhur tuning

### CODE REVIEW

File: autokeras/model_trainer.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        early_stop: An instance of class EarlyStop.
+        metric: It should be a subclass of class autokeras.metric.Metric.
+            In the compute(prediction, target) function, prediction and targets are
+            all numpy arrays converted from the output of the model and the targets packed in the data loaders.
+        verbose: Verbosity mode.
+    """
+
+    def __init__(self, model, train_loader, test_loader, metric, loss_function, verbose):
+        """Init the ModelTrainer with `model`, `x_train`, `y_train`, `x_test`, `y_test`, `verbose`"""
+        super().__init__(loss_function, train_loader, test_loader, metric, verbose)
```
Comment (jhfjhfj1) [inline autokeras/model_trainer.py]: def __init__(**kwargs, model):
    super().__init__(**kwargs)

File: autokeras/unsupervised.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        """
+
+        Args:
+            x_train: A numpy.ndarray instance containing the training data.
+        """
+        pass
+
+    @abstractmethod
+    def generate(self, input_sample):
+        """
```
Comment (jhfjhfj1) [inline autokeras/unsupervised.py]: input_sample=None

File: tests/test_dcgan.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+from autokeras.gan import DCGAN
+from tests.common import clean_dir
+
+
+def mock_train(**kwargs):
+    str(kwargs)
+    return 1, 0
+
+
+@patch('autokeras.search.ModelTrainer.train_model', side_effect=mock_train)
```
Comment (jhfjhfj1) [inline tests/test_dcgan.py]: search -> gan

File: examples/gan.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import numpy as np
+import tensorflow
+from keras.datasets import cifar10
+
+from autokeras.gan import DCGAN
```
Comment (jhfjhfj1) [inline examples/gan.py]: delete the example

**Comment by jhfjhfj1** [comment]
@tl-yang 
Please fix the CI.
Then I will merge.

Thanks.

**Comment by jhfjhfj1** [comment]
@tl-yang I am not sure the selection of GPU part would crash in some special cases. And we need to support manual selection of GPU as well. Thanks.
